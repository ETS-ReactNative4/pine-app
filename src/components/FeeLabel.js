import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';
import PropTypes from 'prop-types';

import { formatPercentage } from '../localization';
import CurrencyLabelContainer from '../containers/CurrencyLabelContainer';
import Bullet from './typography/Bullet';
import StyledText from './StyledText';

const FEE_HIGH_COLOR = '#FF3B30';
const FEE_MEDIUM_COLOR = '#FF9500';

const styles = StyleSheet.create({
  wrapper: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center'
  }
});

export default class FeeLabel extends Component {
  _getPercentage() {
    const { amount, fee } = this.props;
    return fee / Math.abs(amount) * 100;
  }

  _getPercentageColor() {
    const percentage = this._getPercentage();

    if (percentage > 20) {
      return FEE_HIGH_COLOR;
    }

    if (percentage > 5) {
      return FEE_MEDIUM_COLOR;
    }
  }

  _renderFeePercentage() {
    const percentage = this._getPercentage();
    return formatPercentage(percentage);
  }

  render() {
    const { fee, currency, unit, style } = this.props;
    const percentage = this._renderFeePercentage();
    const percentageColor = this._getPercentageColor();

    const percentageStyles = [
      this.props.style,
      percentageColor ? { color: percentageColor } : null
    ];

    return (
      <View style={[styles.wrapper, this.props.style]}>
        <CurrencyLabelContainer
          amountBtc={fee}
          currency={currency}
          unit={unit}
          style={style}
        />
        <Bullet />
        <StyledText style={percentageStyles}>{percentage}</StyledText>
      </View>
    );
  }
}

FeeLabel.propTypes = {
  style: PropTypes.any,
  amount: PropTypes.number,
  fee: PropTypes.number,
  currency: PropTypes.string,
  unit: PropTypes.string
};
