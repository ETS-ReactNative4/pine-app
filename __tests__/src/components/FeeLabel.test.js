import React from 'react';
import renderer from 'react-test-renderer';
import FeeLabel from '../../../src/components/FeeLabel';

jest.mock('../../../src/containers/CurrencyLabelContainer', () => 'CurrencyLabelContainer');

describe('FeeLabel', () => {
  it('renders low fees correctly', () => {
    const tree = renderer.create(
      <FeeLabel amount={1} fee={0.001} unit='BTC' />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders medium fees correctly', () => {
    const tree = renderer.create(
      <FeeLabel amount={2} fee={0.12} unit='BTC' />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders high fees correctly', () => {
    const tree = renderer.create(
      <FeeLabel amount={0.5} fee={0.15} unit='BTC' />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
