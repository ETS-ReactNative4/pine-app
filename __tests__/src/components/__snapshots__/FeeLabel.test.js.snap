// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeeLabel renders high fees correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <CurrencyLabelContainer
    amountBtc={0.15}
    unit="BTC"
  />
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#8A8A8F",
          "borderRadius": 2,
          "height": 3,
          "marginHorizontal": 10,
          "width": 3,
        },
        undefined,
      ]
    }
  />
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Array [
          undefined,
          Object {
            "color": "#FF3B30",
          },
        ],
      ]
    }
  >
    30%
  </Text>
</View>
`;

exports[`FeeLabel renders low fees correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <CurrencyLabelContainer
    amountBtc={0.001}
    unit="BTC"
  />
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#8A8A8F",
          "borderRadius": 2,
          "height": 3,
          "marginHorizontal": 10,
          "width": 3,
        },
        undefined,
      ]
    }
  />
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Array [
          undefined,
          null,
        ],
      ]
    }
  >
    0.1%
  </Text>
</View>
`;

exports[`FeeLabel renders medium fees correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
      },
      undefined,
    ]
  }
>
  <CurrencyLabelContainer
    amountBtc={0.12}
    unit="BTC"
  />
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#8A8A8F",
          "borderRadius": 2,
          "height": 3,
          "marginHorizontal": 10,
          "width": 3,
        },
        undefined,
      ]
    }
  />
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Array [
          undefined,
          Object {
            "color": "#FF9500",
          },
        ],
      ]
    }
  >
    6%
  </Text>
</View>
`;
